### PR TITLE
COLA-MVP-version0.3

### PR DESCRIPTION
**Pull request by ZhuMuMu0216**

### CODE REVIEW

File: cola/ace_generator.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+from .ce_module.ce_models import DiCE
+from .model import Model
+from .data import Data
+from .policy_module import policy
+import numpy as np
+import pandas as pd
+
+class COLA:
+    def __init__(self, ml_model=None, data=None, counterfactual_algorithm=None, policy_name="CF_OTMatch", Avalues_method="max", limited_actions=None):
```
Comment (youlei202) [inline cola/ace_generator.py]: I noticed that you have declared type for other class. I think it is good to do it for all your functions. E.g.

ml_model : Model

File: cola/ace_generator.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        self.ml_model = ml_model
+        self.data = data
+        self.counterfactual_algorithm = counterfactual_algorithm
+        self.limited_actions = limited_actions
+        self.Avalues_method = Avalues_method
+        self.policy_name = policy_name
+        # self.ce_algorithm =None
+        # self.results = self.generate_counterfactual()
+
+    def generate_ace_counterfactual(self):
```
Comment (youlei202) [inline cola/ace_generator.py]: generate_action_constrained_counterfactual sounds better.

It is difficult to get the meaning of "ace"

File: cola/ace_generator.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        
+        # get model
+        self.ml_model_true = self.ml_model.load_model()
+
+        # get ce_algorithm
+        ce_algorithm = self.counterfactualexplanation_algorithm()
```
Comment (youlei202) [inline cola/ace_generator.py]: counterfactual_explanation_algorithm is better

(I usually do not modify the names that you've chosen, unless it sounds really weird by the standard of python development)

File: cola/ace_generator.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        y_counterfactual = self.ml_model_true.predict(x_action_constrained)
+
+        factual = self.return_dataframe(x_factual, y_factual)
+        ce = self.return_dataframe(x_counterfactual, y_counterfactual)
+        ace = self.return_dataframe(x_action_constrained, y_counterfactual)
+        
+        return factual, ce, ace
+
+
+    def counterfactualexplanation_algorithm(self):
```
Comment (youlei202) [inline cola/ace_generator.py]: Later on, you gonna support multiple counterfactual explanations algorithms. I think it is better to explicitly declare as:

def counterfactual_explanation_algorithm(self) -> CounterFactualExplainer:

File: cola/ce_module/base_ce.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from abc import ABC, abstractmethod
+import pandas as pd
+
+class CounterFactualExplainer(ABC):
+    def __init__(self, ml_model, x_factual: pd.DataFrame):
```
Comment (youlei202) [inline cola/ce_module/base_ce.py]: ml_model should also be with a type

File: cola/ace_generator.py (hunk 3 above)
Comment (ZhuMuMu0216) [reply cola/ace_generator.py]: OK, I have already made these changes.

**Review by youlei202 — APPROVED** [review]

**Pull request merged**